lib/inv_pcg_quda.cpp

```diff
@@ -331,49 +349,42 @@ namespace quda
         mat(r, y, x, tmp3); // x is just a temporary here
         r2 = xmyNorm(b, r);
 
-        if (param.deflate && sqrt(r2) < maxr_deflate * param.tol_restart) {
+        if (param.deflate && sqrt(r2) < ru.maxr_deflate * param.tol_restart) {
           // Deflate and accumulate to solution vector
           eig_solve->deflate(y, r, evecs, evals, true);
 
           // Compute r_defl = RHS - A * LHS
           mat(r, y, x, tmp3);
           r2 = blas::xmyNorm(b, r);
 
-          maxr_deflate = sqrt(r2);
+          ru.update_maxr_deflate(r2);
         }
 
         copy(rSloppy, r); // copy r to rSloppy
         zero(xSloppy);
 
-        // break-out check if we have reached the limit of the precision
-        if (sqrt(r2) > r0Norm && updateX) {
-          resIncrease++;
-          resIncreaseTotal++;
-          // reuse r0Norm for this
-          warningQuda(
-            "PCG: new reliable residual norm %e is greater than previous reliable residual norm %e (total #inc %i)",
-            sqrt(r2), r0Norm, resIncreaseTotal);
-
-          if (resIncrease > maxResIncrease or resIncreaseTotal > maxResIncreaseTotal) break;
+        bool L2breakdown = false;
+        double L2breakdown_eps = 0;
+        if (ru.reliable_break(r2, stop, L2breakdown, L2breakdown_eps)) { break; }
 
-        } else {
-          resIncrease = 0;
-        }
+        ru.update_norm(r2, y);
 
-        rNorm = sqrt(r2);
-        maxrr = rNorm;
-        maxrx = rNorm;
-        r0Norm = rNorm;
-        ++rUpdate;
+        ru.reset(r2);
 
         if (K) {
+          // can fuse these two kernels
+          r_new_Minvr_old = reDotProduct(rSloppy, *minvrSloppy);
           *rPre = rSloppy;
+
+          pushVerbosity(param.verbosity_precondition);
           (*K)(*minvrPre, *rPre);
-          *minvrSloppy = *minvrPre;
+          popVerbosity();
 
+          // can fuse these two kernels
+          *minvrSloppy = *minvrPre;
           rMinvr = reDotProduct(rSloppy, *minvrSloppy);
-          beta = rMinvr / rMinvr_old;
 
+          beta = (rMinvr - r_new_Minvr_old) / rMinvr_old;
```

Comment:
I'm guessing this is because we didn't have Polak-Ribière in PCG before? (Which means I misread PCG in the past!)

Reply:
I believe that is the case.

Reply:
@weinbe2 Looking at this again, we did have Polak-Ribière in PCG previously, but only when reliable update is not triggered. The code path here is the one where reliable update is performed.